### ISSUE TITLE
model robustness tests

### ISSUE THREAD

**Issue by changhoonhahn** [galactic-ai/haloflow#10]
Test robustness of NPEs by comparing the M* and Mh posteriors for **NPE trained on simulation X, tested on simulation Y**.

**Comment by Nikhil0504** [comment]
I started writing some code to do the cross-validation based on the validation scripts we already have. See #13 for more information.

Here are a few results upon training with `v1` datasets. All the plots are in `/xdisk/chhahn/chhahn/haloflow/hf2/npe`. 
It will be as `h2.v1.{train_sim}.{obs}.tarp_crossvalid.png` and `h2.v1.{train_sim}.{obs}.rank_crossvalid.png`

## Trained with SIMBA 100 data

![Image](https://github.com/user-attachments/assets/e62ff788-f7bc-4980-8cb1-b73b2ca0e175)

![Image](https://github.com/user-attachments/assets/d4bdd8e3-6bd8-4fec-9a5f-695c8ba97532)

## Trained with Eagle 100 data

![Image](https://github.com/user-attachments/assets/6256be4c-789d-4355-b734-3d1ac8cc9138)

![Image](https://github.com/user-attachments/assets/0faf57cc-80d3-4895-8fed-e72f3b3ed2bd)

## Trained with TNG 100 data

![Image](https://github.com/user-attachments/assets/6f7e5a7e-cc1f-4f04-ad26-4abb4b77444f)

![Image](https://github.com/user-attachments/assets/b7f7da9c-b48c-4517-a521-58a9fdc181ee)

**Comment by changhoonhahn** [comment]
Great figures. Seems TNG100 - Simba100 are somewhat consistent but the rest are not robust at all. 

A few questions: 
1. Are these with the prior correction I mentioned during our last meeting? 
2. Rank 1 and 2 correspond to M* and Mh? 
3. Have you looked at a true vs predicted M* and Mh comparison?

**Comment by Nikhil0504** [comment]
> 1. Are these with the prior correction I mentioned during our last meeting?

No, haven't implemented that yet. I wasn't sure for which sims we had to use the prior correction.

> 3. Rank 1 and 2 correspond to M* and Mh?

Yes.

> 5. Have you looked at a true vs predicted M* and Mh comparison?

Not yet, will implement the scripts today to have a look.

**Comment by changhoonhahn** [comment]
Prior corrections should be implemented for all simulations. 

For inferring $M_*$ and $M_h$, we ideally want uniform uninformative priors but we have to contend with the SMF and HMF (i.e. there are more low mass galaxies/halos in the sample). So each NPE has to be corrected for the prior that's set by $p(M_*, M_h)$ of the training dataset.

**Comment by Nikhil0504** [comment]
Here is a preview of all the models correlated with each other. Blue points are unweighted and red points are weighted. I'll upload the M_* scatter plots soon.
![Image](https://github.com/user-attachments/assets/409770bb-8df5-4f36-b52c-2d9370794c92)

**Comment by Nikhil0504** [comment]
Stellar mass

![Image](https://github.com/user-attachments/assets/7c34c5d7-24ff-4601-9ae1-bf750fe27001)

**Comment by Nikhil0504** [comment]
Hello @cbottrell, I also trained the simulations to use the rest of the features from the catalogs you made. 
They are using all these parameters for each filter:
https://github.com/galactic-ai/haloflow/blob/0dc7e6a1d37e7fe252a678ded766e5eea26a9f65/src/haloflow/data.py#L30-L32

<img width="805" alt="Image" src="https://github.com/user-attachments/assets/a2d51ac8-b818-41fe-a609-77ac85e2e469" />

<img width="811" alt="Image" src="https://github.com/user-attachments/assets/995e9558-46df-4bd6-ab77-c354ec649fe9" />

Interestingly, there seems to be a tighter correlation when we add these extra features when we cross-validate with different simulations. I am curious to know your thoughts on this.

Here are the coverage plots for all the runs:
![Image](https://github.com/user-attachments/assets/62db3c88-c6c4-4219-8b0e-244b0be8bf7b)

@changhoonhahn are there any other tests we should do for this?

**Comment by changhoonhahn** [comment]
The tests are sufficient for now. The overall robustness even with the extra morphological properties is certainly interesting. I initially expected that as the constraints get tighter, we'll find much worse robustness.

**Comment by cbottrell** [comment]
@Nikhil0504 @changhoonhahn The correlations sure do look better with cross-validation. Do you mean here that, e.g., if you train a model on TNG100 training data and apply it to EAGLE test data, it does better if your validation set is from EAGLE? If so, that's not much of a surprise. I would expect that such a model would also have slightly poorer accuracy on test data taken from TNG100. 

Each set of baryon physics will produce slight variations on the galaxies living in halos of roughly the same mass. The correlations between halo mass and galaxy properties will differ slightly. By cross-validating, you are essentially preventing the model from over-fitting to the correlations implicit to the baryon physics used in training. That would be my guess. 

What happens when you cross-validate on another simulation but apply the model to test data drawn from the original? E.g. You train on TNG, validate on Eagle, and test on TNG? How does the model perform relative to self-validation? 

Nice work!

**Comment by Nikhil0504** [comment]
@cbottrell, right now I train and validate on the same simulation and use those models to further independently test the models in scatter plots shown above.

For example: I use `obs='mags'` and `sim='TNG50` to train the NPEs.
https://github.com/galactic-ai/haloflow/blob/4039942ea8190e4ae8c25a396df8e8149893b9a0/bin/npe/npe.py#L29
Here is the training loop:
https://github.com/galactic-ai/haloflow/blob/4039942ea8190e4ae8c25a396df8e8149893b9a0/bin/npe/npe.py#L67-L87

Here is my testing script: https://github.com/galactic-ai/haloflow/blob/4039942ea8190e4ae8c25a396df8e8149893b9a0/bin/npe/cross_validation_scatter_plots.ipynb

Which is inherently does this and samples the NDEs accordingly:
https://github.com/galactic-ai/haloflow/blob/4039942ea8190e4ae8c25a396df8e8149893b9a0/src/haloflow/npe/valid.py#L22-L29

**Comment by Nikhil0504** [comment]
```python
 for i, train_sim in enumerate(sims):
     for j, test_sim in enumerate(sims):
         if train_sim == test_sim:
             continue
         MODEL_NAME = f'dann_model_to_{test_sim}_{obs}'
         fp = dat_dir + f'hf2/dann/models/{MODEL_NAME}.pt'
 
         ax[j, i], _, _ = P.plot_true_pred(ax[j, i],
                                           obs,
                                           train_sim,
                                           obs,
                                           test_sim,
                                           'cpu',
                                           fmt='.b',
                                           with_dann=True,
                                           data_dir=data_dir,
                                           fp=fp,
                                           mass='stellar',
                                           version=2
                                           )
 
 
         ax[j, i], _, _ = P.plot_true_pred(ax[j, i],
                                           obs,
                                           train_sim,
                                           obs,
                                           test_sim,
                                           'cpu',
                                           fmt='.r',
                                           with_dann=True,
                                           data_dir=data_dir,
                                           fp=fp,
                                           mass='stellar',
                                           version=2,
                                           use_weights=True,
                                           )
         ax[j, i].legend()
         # remove x-axis labels  
         if j != len(sims)-1:
             ax[j, i].set_xticklabels([])
             ax[j, i].set_xlabel('')
         # remove y-axis labels
         if i != 0:
             ax[j, i].set_yticklabels([])
             ax[j, i].set_ylabel('')
 
 plt.tight_layout()
 plt.savefig('halo_scatter_plot_npe_test.png', dpi=150)
 plt.show()
```
The error bars are too big,
![Image](https://github.com/user-attachments/assets/2ffed7b4-9caa-4862-859b-04f9594d6949)

**Comment by Nikhil0504** [comment]
with `mags_morph`

![Image](https://github.com/user-attachments/assets/04280854-4625-4ca6-9695-44d05c416efe)

**Comment by Nikhil0504** [comment]
![Image](https://github.com/user-attachments/assets/011f21fe-008f-47f4-a62e-a45e9b2cc7fe)
With `mags_morph_extra`, a lot lot better

**Comment by Nikhil0504** [comment]
![Image](https://github.com/user-attachments/assets/f9b8de8c-c412-4e2c-957a-07ef8cbfd348)
Blue is without DANN and red is with DANN

**Comment by Nikhil0504** [comment]
@changhoonhahn 
![Image](https://github.com/user-attachments/assets/95ddefb6-3ba9-4778-8512-7b404d7d3c48)
Blue is without DANN and red is with DANN for `mags_morph_extra`

**Comment by changhoonhahn** [comment]
We're currently treating `tng50` as a separate simulation. This may be an issue because `tng50` has higher resolution than the rest of the simulation, so its stellar mass-to-halo mass relation may be significantly different than the rest. 

Rather than treating `tng50` and `tng100` as separate simulations, let's combine them with the corrections as `tng_all` and repeat the DANN posterior comparisons.

**Comment by cbottrell** [comment]
![Image](https://github.com/user-attachments/assets/35a5d337-ad5f-4892-84d4-0af019317190)

The TNG SHMR changes with resolution. Higher resolution sims (TNG50-1) have higher stellar masses per unit halo mass than lower resolution sims (e.g. TNG50-3). I've given you tables with corrected halo masses and original halo masses. This correction is simply a linearly interpolated scaling of stellar masses in TNG50-1 to match the SHMR of TNG100 (the volume for which the baryon physics were calibrated to match the z=0 stellar mass function). See Figure above.

![Image](https://github.com/user-attachments/assets/8e97eb9e-59d5-4134-a9b6-f79689731580)

This figure shows the SHMRs of each simulation, including the rescaled TNG50-1 (called rTNG50-1).
